File: zenmap/windows/native/Services/XmlParsing.cs
using System.Text.RegularExpressions;
using System.Xml.Linq;
using Zenmap.Windows.Models;

namespace Zenmap.Windows.Services;

public static class XmlParsing
{
    private static readonly Regex CompleteHostRegex = new(
        @"<host\b[^>]*>.*?</host>",
        RegexOptions.Singleline | RegexOptions.IgnoreCase | RegexOptions.Compiled);

    private static readonly Regex CompleteHostHintRegex = new(
        @"<hosthint\b[^>]*>.*?</hosthint>",
        RegexOptions.Singleline | RegexOptions.IgnoreCase | RegexOptions.Compiled);

    private static readonly Regex DiscoveredOpenPortRegex = new(
        @"Discovered open port (\d+)/(tcp|udp|sctp) on (\S+)",
        RegexOptions.IgnoreCase | RegexOptions.Compiled);

    private static readonly Regex ScanReportRegex = new(
        @"Nmap scan report for (?:(\S+) \((\d[^)]*)\)|(\S+))",
        RegexOptions.IgnoreCase | RegexOptions.Compiled);

    public static IReadOnlyList<ScannedHost> ParseNmapXml(string path)
    {
        if (!File.Exists(path))
        {
            return Array.Empty<ScannedHost>();
        }

        string text;
        try
        {
            text = File.ReadAllText(path);
        }
        catch (Exception)
        {
            return Array.Empty<ScannedHost>();
        }

        if (string.IsNullOrWhiteSpace(text))
        {
            return Array.Empty<ScannedHost>();
        }

        try
        {
            var document = XDocument.Parse(text, LoadOptions.None);
            var finished = document.Root?
                .Elements("host")
                .Select(ParseHost)
                .Where(host => host is not null)
                .Cast<ScannedHost>()
                .ToArray() ?? Array.Empty<ScannedHost>();
            var hints = document.Root?
                .Elements("hosthint")
                .Select(ParseHost)
                .Where(host => host is not null)
                .Cast<ScannedHost>()
                .ToArray() ?? Array.Empty<ScannedHost>();
            return MergeScanHosts(finished, hints);
        }
        catch (Exception)
        {
            return MergeScanHosts(
                ParseCompleteElements(text, CompleteHostRegex),
                ParseCompleteElements(text, CompleteHostHintRegex));
        }
    }

    public static IReadOnlyList<ScannedHost> ParseLiveOutputHosts(string outputText)
    {
        var hostsByAddress = new Dictionary<string, ScannedHost>(StringComparer.Ordinal);

        foreach (Match match in ScanReportRegex.Matches(outputText))
        {
            var hostname = match.Groups[1].Success ? match.Groups[1].Value : "";
            var parentheticalIp = match.Groups[2].Success ? match.Groups[2].Value : "";
            var bareTarget = match.Groups[3].Success ? match.Groups[3].Value : "";
            var address = string.IsNullOrWhiteSpace(parentheticalIp) ? bareTarget : parentheticalIp;
            if (string.IsNullOrWhiteSpace(address))
            {
                continue;
            }

            if (hostsByAddress.TryGetValue(address, out var existing))
            {
                if (string.IsNullOrWhiteSpace(existing.Hostname) && !string.IsNullOrWhiteSpace(hostname))
                {
                    existing.Hostname = hostname;
                }
            }
            else
            {
                hostsByAddress[address] = new ScannedHost
                {
                    Address = address,
                    Hostname = hostname,
                    Status = "up",
                };
            }
        }

        foreach (Match match in DiscoveredOpenPortRegex.Matches(outputText))
        {
            var portNumber = match.Groups[1].Value;
            var protocolName = match.Groups[2].Value.ToLowerInvariant();
            var address = match.Groups[3].Value;
            if (!hostsByAddress.TryGetValue(address, out var host))
            {
                host = new ScannedHost { Address = address, Status = "up" };
                hostsByAddress[address] = host;
            }

            if (host.Ports.Any(port =>
                    string.Equals(port.ProtocolName, protocolName, StringComparison.OrdinalIgnoreCase) &&
                    port.PortNumber == portNumber))
            {
                continue;
            }

            host.Ports.Add(new ScannedPort
            {
                HostAddress = address,
                ProtocolName = protocolName,
                PortNumber = portNumber,
                State = "open",
            });
        }

        return hostsByAddress.Values.ToArray();
    }

    public static IReadOnlyList<ScannedHost> MergeScanHosts(
        IReadOnlyList<ScannedHost> primary,
        IReadOnlyList<ScannedHost> secondary)
    {
        var merged = new Dictionary<string, ScannedHost>(StringComparer.Ordinal);
        var order = new List<string>();

        foreach (var source in new[] { primary, secondary })
        {
            foreach (var host in source)
            {
                var key = string.IsNullOrWhiteSpace(host.Address) ? host.Hostname : host.Address;
                if (string.IsNullOrWhiteSpace(key))
                {
                    continue;
                }

                if (!merged.TryGetValue(key, out var existing))
                {
                    merged[key] = CloneHost(host);
                    order.Add(key);
                    continue;
                }

                if (string.IsNullOrWhiteSpace(existing.Hostname) && !string.IsNullOrWhiteSpace(host.Hostname))
                {
                    existing.Hostname = host.Hostname;
                }

                if ((string.IsNullOrWhiteSpace(existing.Status) || existing.Status == "unknown") &&
                    !string.IsNullOrWhiteSpace(host.Status))
                {
                    existing.Status = host.Status;
                }

                existing.Ports = MergePorts(existing.Ports, host.Ports);
            }
        }

        return order.Select(key => merged[key]).ToArray();
    }

    public static string HostsFingerprint(IReadOnlyList<ScannedHost> hosts) =>
        string.Join(
            '\n',
            hosts.Select(host =>
                $"{host.Address}|{host.Hostname}|{host.Status}|" +
                string.Join(
                    ';',
                    host.Ports.Select(port =>
                        $"{port.ProtocolName}/{port.PortNumber}/{port.State}/{port.ServiceName}/{port.Product}/{port.Version}"))));

    private static IReadOnlyList<ScannedHost> ParseCompleteElements(string text, Regex pattern)
    {
        var hosts = new List<ScannedHost>();
        foreach (Match match in pattern.Matches(text))
        {
            try
            {
                var host = ParseHost(XElement.Parse(match.Value));
                if (host is not null)
                {
                    hosts.Add(host);
                }
            }
            catch (Exception)
            {
            }
        }

        return hosts;
    }

    private static List<ScannedPort> MergePorts(IEnumerable<ScannedPort> primary, IEnumerable<ScannedPort> secondary)
    {
        var merged = new Dictionary<string, ScannedPort>(StringComparer.OrdinalIgnoreCase);
        var order = new List<string>();

        foreach (var source in new[] { primary, secondary })
        {
            foreach (var port in source)
            {
                var key = $"{port.ProtocolName}/{port.PortNumber}";
                if (!merged.TryGetValue(key, out var existing))
                {
                    merged[key] = ClonePort(port);
                    order.Add(key);
                    continue;
                }

                if (string.IsNullOrWhiteSpace(existing.ServiceName)) existing.ServiceName = port.ServiceName;
                if (string.IsNullOrWhiteSpace(existing.Product)) existing.Product = port.Product;
                if (string.IsNullOrWhiteSpace(existing.Version)) existing.Version = port.Version;
                if (string.IsNullOrWhiteSpace(existing.ExtraInfo)) existing.ExtraInfo = port.ExtraInfo;
                if (string.IsNullOrWhiteSpace(existing.State) || existing.State == "unknown") existing.State = port.State;
            }
        }

        return order.Select(key => merged[key]).ToList();
    }

    private static ScannedHost CloneHost(ScannedHost host) => new()
    {
        Address = host.Address,
        Hostname = host.Hostname,
        Status = host.Status,
        Ports = host.Ports.Select(ClonePort).ToList(),
    };

    private static ScannedPort ClonePort(ScannedPort port) => new()
    {
        HostAddress = port.HostAddress,
        ProtocolName = port.ProtocolName,
        PortNumber = port.PortNumber,
        State = port.State,
        ServiceName = port.ServiceName,
        Product = port.Product,
        Version = port.Version,
        ExtraInfo = port.ExtraInfo,
    };

    private static ScannedHost? ParseHost(XElement hostElement)
    {
        var status = hostElement.Element("status")?.Attribute("state")?.Value ?? "unknown";

        var address = "";
        foreach (var addressElement in hostElement.Elements("address"))
        {
            var addrType = addressElement.Attribute("addrtype")?.Value ?? "";
            var candidate = addressElement.Attribute("addr")?.Value ?? "";
            if (string.IsNullOrWhiteSpace(candidate))
            {
                continue;
            }

            if (string.IsNullOrWhiteSpace(address) || addrType is "ipv4" or "ipv6")
            {
                address = candidate;
                if (addrType is "ipv4" or "ipv6")
                {
                    break;
                }
            }
        }

        var hostname = hostElement.Element("hostnames")?
            .Element("hostname")?
            .Attribute("name")?
            .Value ?? "";

        var ports = hostElement.Element("ports")?
            .Elements("port")
            .Select(port => ParsePort(port, address))
            .Where(port => port is not null)
            .Cast<ScannedPort>()
            .ToArray() ?? Array.Empty<ScannedPort>();

        if (string.IsNullOrWhiteSpace(address) && string.IsNullOrWhiteSpace(hostname))
        {
            return null;
        }

        return new ScannedHost
        {
            Address = address,
            Hostname = hostname,
            Status = status,
            Ports = ports.ToList(),
        };
    }

    private static ScannedPort? ParsePort(XElement portElement, string hostAddress)
    {
        var state = portElement.Element("state")?.Attribute("state")?.Value ?? "unknown";
        var service = portElement.Element("service");

        return new ScannedPort
        {
            HostAddress = hostAddress,
            ProtocolName = portElement.Attribute("protocol")?.Value ?? "",
            PortNumber = portElement.Attribute("portid")?.Value ?? "",
            State = state,
            ServiceName = service?.Attribute("name")?.Value ?? "",
            Product = service?.Attribute("product")?.Value ?? "",
            Version = service?.Attribute("version")?.Value ?? "",
            ExtraInfo = service?.Attribute("extrainfo")?.Value ?? "",
        };
    }
}

File: zenmap/linux/native/xml_parsing.py
"""Parse Nmap XML output into platform-neutral host models."""

from __future__ import annotations

import re
import xml.etree.ElementTree as ET
from pathlib import Path

from .models import ScannedHost, ScannedPort

# ``<host\b`` does not match ``<hosthint>``.
_COMPLETE_HOST_RE = re.compile(r"<host\b[^>]*>.*?</host>", re.DOTALL | re.IGNORECASE)
_COMPLETE_HOSTHINT_RE = re.compile(
    r"<hosthint\b[^>]*>.*?</hosthint>",
    re.DOTALL | re.IGNORECASE,
)
_DISCOVERED_OPEN_PORT_RE = re.compile(
    r"Discovered open port (\d+)/(tcp|udp|sctp) on (\S+)",
    re.IGNORECASE,
)
_SCAN_REPORT_RE = re.compile(
    r"Nmap scan report for (?:(\S+) \((\d[^)]*)\)|(\S+))",
    re.IGNORECASE,
)


def parse_nmap_xml(path: str | Path) -> list[ScannedHost]:
    """Parse Nmap XML, including incomplete mid-scan files.

    Mid-scan XML often contains finished ``<hosthint>`` discovery records before
    any closed ``<host>`` element. Full-document parses fail until the scan ends,
    so fall back to extracting complete host/hosthint fragments.
    """
    xml_path = Path(path)
    if not xml_path.is_file():
        return []

    try:
        text = xml_path.read_text(encoding="utf-8", errors="replace")
    except OSError:
        return []

    if not text.strip():
        return []

    try:
        root = ET.fromstring(text)
    except ET.ParseError:
        return _merge_hosts(
            _parse_complete_elements(text, _COMPLETE_HOST_RE),
            _parse_complete_elements(text, _COMPLETE_HOSTHINT_RE),
        )

    finished = [
        host
        for host_element in root.findall("host")
        if (host := _parse_host(host_element)) is not None
    ]
    hints = [
        host
        for hint_element in root.findall("hosthint")
        if (host := _parse_host(hint_element)) is not None
    ]
    return _merge_hosts(finished, hints)


def parse_live_output_hosts(output_text: str) -> list[ScannedHost]:
    """Build provisional hosts/ports from verbose nmap stdout."""
    hosts_by_address: dict[str, ScannedHost] = {}

    for match in _SCAN_REPORT_RE.finditer(output_text):
        hostname, parenthetical_ip, bare_target = match.groups()
        if parenthetical_ip:
            address = parenthetical_ip
            host_name = hostname or ""
        else:
            address = bare_target or ""
            host_name = ""
        if not address:
            continue
        host = hosts_by_address.get(address)
        if host is None:
            hosts_by_address[address] = ScannedHost(
                address=address,
                hostname=host_name,
                status="up",
                ports=[],
            )
        elif host_name and not host.hostname:
            host.hostname = host_name

    for match in _DISCOVERED_OPEN_PORT_RE.finditer(output_text):
        port_number, protocol_name, address = match.groups()
        host = hosts_by_address.get(address)
        if host is None:
            host = ScannedHost(address=address, hostname="", status="up", ports=[])
            hosts_by_address[address] = host
        key = (protocol_name.lower(), port_number)
        if any(
            (port.protocol_name.lower(), port.port_number) == key for port in host.ports
        ):
            continue
        host.ports.append(
            ScannedPort(
                host_address=address,
                protocol_name=protocol_name.lower(),
                port_number=port_number,
                state="open",
            )
        )

    return list(hosts_by_address.values())


def merge_scan_hosts(
    xml_hosts: list[ScannedHost],
    live_output_hosts: list[ScannedHost],
) -> list[ScannedHost]:
    """Prefer richer XML host records, then fill gaps from live stdout."""
    return _merge_hosts(xml_hosts, live_output_hosts)


def hosts_fingerprint(hosts: list[ScannedHost]) -> tuple:
    """Stable fingerprint used to skip redundant live UI refreshes."""
    return tuple(
        (
            host.address,
            host.hostname,
            host.status,
            tuple(
                (
                    port.protocol_name,
                    port.port_number,
                    port.state,
                    port.service_name,
                    port.product,
                    port.version,
                )
                for port in host.ports
            ),
        )
        for host in hosts
    )


def _parse_complete_elements(text: str, pattern: re.Pattern[str]) -> list[ScannedHost]:
    hosts: list[ScannedHost] = []
    for match in pattern.finditer(text):
        try:
            host_element = ET.fromstring(match.group(0))
        except ET.ParseError:
            continue
        host = _parse_host(host_element)
        if host is not None:
            hosts.append(host)
    return hosts


def _merge_hosts(
    primary: list[ScannedHost],
    secondary: list[ScannedHost],
) -> list[ScannedHost]:
    merged: dict[str, ScannedHost] = {}
    order: list[str] = []

    for source in (primary, secondary):
        for host in source:
            key = host.address or host.hostname
            if not key:
                continue
            existing = merged.get(key)
            if existing is None:
                merged[key] = ScannedHost(
                    address=host.address,
                    hostname=host.hostname,
                    status=host.status,
                    ports=list(host.ports),
                )
                order.append(key)
                continue
            if host.hostname and not existing.hostname:
                existing.hostname = host.hostname
            if host.status and existing.status in {"", "unknown"}:
                existing.status = host.status
            existing.ports = _merge_ports(existing.ports, host.ports)

    return [merged[key] for key in order]


def _merge_ports(
    primary: list[ScannedPort],
    secondary: list[ScannedPort],
) -> list[ScannedPort]:
    merged: dict[tuple[str, str], ScannedPort] = {}
    order: list[tuple[str, str]] = []

    for source in (primary, secondary):
        for port in source:
            key = (port.protocol_name.lower(), port.port_number)
            existing = merged.get(key)
            if existing is None:
                merged[key] = ScannedPort(
                    host_address=port.host_address,
                    protocol_name=port.protocol_name,
                    port_number=port.port_number,
                    state=port.state,
                    service_name=port.service_name,
                    product=port.product,
                    version=port.version,
                    extra_info=port.extra_info,
                )
                order.append(key)
                continue
            if port.service_name and not existing.service_name:
                existing.service_name = port.service_name
            if port.product and not existing.product:
                existing.product = port.product
            if port.version and not existing.version:
                existing.version = port.version
            if port.extra_info and not existing.extra_info:
                existing.extra_info = port.extra_info
            if port.state and existing.state in {"", "unknown"}:
                existing.state = port.state

    return [merged[key] for key in order]


def _parse_host(host_element: ET.Element) -> ScannedHost | None:
    status_element = host_element.find("status")
    status = status_element.get("state", "unknown") if status_element is not None else "unknown"

    address = ""
    for address_element in host_element.findall("address"):
        addr_type = address_element.get("addrtype", "")
        candidate = address_element.get("addr", "")
        if not candidate:
            continue
        if not address or addr_type in {"ipv4", "ipv6"}:
            address = candidate
            if addr_type in {"ipv4", "ipv6"}:
                break

    hostname = ""
    hostnames_element = host_element.find("hostnames")
    if hostnames_element is not None:
        hostname_element = hostnames_element.find("hostname")
        if hostname_element is not None:
            hostname = hostname_element.get("name", "")

    ports: list[ScannedPort] = []
    ports_element = host_element.find("ports")
    if ports_element is not None:
        for port_element in ports_element.findall("port"):
            port = _parse_port(port_element, address)
            if port is not None:
                ports.append(port)

    if not address and not hostname:
        return None

    return ScannedHost(address=address, hostname=hostname, status=status, ports=ports)


def _parse_port(port_element: ET.Element, host_address: str) -> ScannedPort | None:
    state_element = port_element.find("state")
    service_element = port_element.find("service")

    return ScannedPort(
        host_address=host_address,
        protocol_name=port_element.get("protocol", ""),
        port_number=port_element.get("portid", ""),
        state=state_element.get("state", "unknown") if state_element is not None else "unknown",
        service_name=service_element.get("name", "") if service_element is not None else "",
        product=service_element.get("product", "") if service_element is not None else "",
        version=service_element.get("version", "") if service_element is not None else "",
        extra_info=service_element.get("extrainfo", "") if service_element is not None else "",
    )
